src/Map.ts

```diff
@@ -1190,4 +1208,209 @@ export class Map extends maplibregl.Map {
+      throw new Error(
+        "Creating a polyline layer requires an existing .sourceId or a valid .data property",
+      );
+    }
+
+    // We are going to evaluate the content of .data, if provided
+    let data = options.data as any;
+    let tmpData = null;
+
+    if (typeof data === "string") {
+      // if options.data exists and is a uuid string, we consider that it points to a MapTiler Dataset
+      if (isUUID(data)) {
+        data = `https://api.maptiler.com/data/${options.data}/features.json?key=${config.apiKey}`;
+      }
+
+      // options.data could be a url to a .gpx file
+      else if (data.split(".").pop()?.toLowerCase().trim() === "gpx") {
+        // fetch the file
+        const res = await fetch(data, fetchOptions);
+        const gpxStr = await res.text();
+        // Convert it to geojson. Will throw is invalid GPX content
+        data = gpx(gpxStr) as FeatureCollection;
+      }
+
+      // options.data could be a url to a .kml file
+      else if (data.split(".").pop()?.toLowerCase().trim() === "kml") {
+        // fetch the file
+        const res = await fetch(data, fetchOptions);
+        const kmlStr = await res.text();
+        // Convert it to geojson. Will throw is invalid GPX content
+        data = kml(kmlStr) as FeatureCollection;
+      } else if (
+        (tmpData = jsonParseNoThrow(data) ?? gpxOrKml(data)) !== null
+      ) {
+        // From this point, we consider that the string content provided could
+        // be the string content of one of the compatible format (GeoJSON, KML, GPX)
+        data = tmpData;
+      }
+
+      if (!data) {
+        throw new Error(
+          "Polyline data was provided as string but is incompatible with valid formats.",
+        );
+      }
+    }
+
+    // Data was provided as a non-string but it's not a valid GeoJSON either => throw
+    else if (data && !geojsonValidation.valid(data)) {
+      throw new Error(
+        "Polyline data was provided as an object but object is not of a valid GeoJSON format",
+      );
+    }
+
+    return this.addGeoJSONPolyline({
+      ...options,
+      data,
+    });
+  }
+
+  /**
+   * Add a polyline witgh optional outline from a GeoJSON object
+   * @param data
+   * @param options
+   * @returns
+   */
+  private addGeoJSONPolyline(
+    // this Feature collection is expected to contain on LineStrings and MultilLinestrings
+    options: PolylineLayerOptions,
+  ): {
+    polylineLayerId: string;
+    polylineOutlineLayerId: string | null;
+    polylineSourceId: string;
+  } {
+    if (options.layerId && this.getLayer(options.layerId)) {
+      throw new Error(
+        `A layer already exists with the layer id: ${options.layerId}`,
+      );
+    }
+
+    const sourceId = options.sourceId ?? generateRandomSourceName();
+    const layerId = options.layerId ?? generateRandomLayerName();
+
+    const retunedInfo = {
+      polylineLayerId: layerId,
+      polylineOutlineLayerId: "",
```

Comment:
If you don't want to remove null from the comment above, then I would suggest setting this to null rather than an empty string.

Reply:
Yeah i'm gonna keep the string and remove the null, it's more consitent and playing with multiple types for the same variable is always a bit annoying